File: packages/anchor/programs/covantic/src/instructions/declare_governance_baseline.rs
use anchor_lang::prelude::*;
use anchor_spl::associated_token::get_associated_token_address;
use anchor_spl::token::{Mint, TokenAccount};

use crate::constants::*;
use crate::errors::CovanticError;
use crate::events::GovernanceBaselineDeclared;
use crate::instructions::checkpoint_authority::observe;
use crate::state::{BaselineView, GovernanceBaseline, InsurancePolicy, ProtocolConfig};

/// What the holder commits to about who may control their agent.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct GovernanceManifest {
    /// Expected owner of the covered token account. Normally the agent.
    pub token_owner: Pubkey,
    pub expected_delegate: Option<Pubkey>,
    pub expected_close_authority: Option<Pubkey>,
    /// Must be `None`. See `GovernanceBaseline::program_upgrade_authority`.
    pub program_upgrade_authority: Option<Pubkey>,
    /// Must be `None`. See `GovernanceBaseline::controller`.
    pub controller: Option<Pubkey>,
    /// Must be zero.
    pub controller_min_threshold: u16,
    /// Extra addresses permitted to hold control. Capped at
    /// {@link MAX_GOVERNANCE_EXTRA_AUTHORITIES}.
    pub extra_authorities: Vec<Pubkey>,
    /// sha256 of the off-chain manifest covering anything richer.
    pub manifest_hash: [u8; 32],
}

/// Declare — or refresh — the authority set that is legitimate for an agent.
///
/// **Holder-signed, and that is the entire point.** Every other evidence
/// mechanism in this protocol has the operator asserting something about the
/// policyholder's situation. Here the policyholder asserts it themselves, in
/// advance, and the program later checks reality against their own statement.
/// It turns "was this authorised?" from an inference into a set membership
/// test.
///
/// Three rules make the statement worth anything:
///
/// **It matures on a delay.** `effective_at` is `GOVERNANCE_BASELINE_DELAY`
/// in the future, and `verify_and_payout_governance` refuses a baseline that
/// had not matured before the pre-incident reading it settles against.
/// Without that, a stolen holder key would declare a convenient set and claim
/// against it in the next instruction. With it, a fraudulent claim has to be
/// set up on chain, in public, an hour ahead.
///
/// **It is bound to the account as it stands.** The covered account is read
/// here, and the declaration is refused unless its current owner, delegate
/// and close authority are all inside the set being declared, and the account
/// is not frozen. A holder who has already handed the account to an outside
/// key cannot then describe the agent as its rightful owner and, an hour
/// later, present the pre-existing arrangement as a takeover.
///
/// **The previous declaration is retained whole.** A refresh keeps every
/// role, wildcard, hash and timestamp of the declaration it replaces, so a
/// rotation landing between the takeover and the claim cannot erase the only
/// usable "before" — the same hole `PolicyBalanceCheckpoint.prev_*` closes.
///
/// What the program does *not* do is interpret `manifest_hash`. It cannot
/// decode a Squads config or an allowed-signer list, and pretending it could
/// would be worse than committing to the hash and leaving the richer check to
/// a reader who can perform it. For the same reason it refuses a manifest
/// that names a program upgrade authority or a controller: the checkpoint
/// reads a token account and can observe neither, so accepting them would
/// sell coverage the settlement path has no way to adjudicate.
pub fn declare_governance_baseline_handler(
    ctx: Context<DeclareGovernanceBaseline>,
    manifest: GovernanceManifest,
) -> Result<()> {
    let clock = Clock::get()?;
    let now = clock.unix_timestamp;
    let policy = &ctx.accounts.policy;

    policy.assert_readable()?;
    require!(
        policy.state == InsurancePolicy::STATE_ACTIVE,
        CovanticError::PolicyNotActive
    );
    require!(
        manifest.token_owner != Pubkey::default(),
        CovanticError::InvalidGovernanceBaseline
    );
    // Roles the program cannot observe are refused rather than recorded. A
    // baseline that lists them reads, to a holder, as coverage; on chain it
    // is a field nothing ever compares.
    require!(
        manifest.program_upgrade_authority.is_none()
            && manifest.controller.is_none()
            && manifest.controller_min_threshold == 0,
        CovanticError::UnsupportedGovernanceRole
    );
    // `None` is how an optional role says "nobody". `Some(zero)` is malformed
    // state that a default-keyed candidate would match against.
    require!(
        manifest.expected_delegate != Some(Pubkey::default())
            && manifest.expected_close_authority != Some(Pubkey::default()),
        CovanticError::InvalidGovernanceBaseline
    );
    require!(
        manifest.extra_authorities.len() <= MAX_GOVERNANCE_EXTRA_AUTHORITIES,
        CovanticError::TooManyGovernanceAuthorities
    );
    // The zero pubkey must never end up in the permitted set: every unset
    // slot in the fixed array is zero, and an attacker who could get control
    // assigned to it would land inside the declared set for free.
    require!(
        !manifest.extra_authorities.contains(&Pubkey::default()),
        CovanticError::InvalidGovernanceBaseline
    );

    let mut extra_authorities = [Pubkey::default(); MAX_GOVERNANCE_EXTRA_AUTHORITIES];
    for (slot, key) in manifest.extra_authorities.iter().enumerate() {
        extra_authorities[slot] = *key;
    }
    let extra_authority_count = manifest.extra_authorities.len() as u8;

    // Bind the declaration to the account it describes. What is declared
    // has to be true *now*; a declaration that describes a state the account
    // is not in is not a baseline, it is a story.
    let declared = BaselineView {
        token_owner: manifest.token_owner,
        expected_delegate: manifest.expected_delegate,
        expected_close_authority: manifest.expected_close_authority,
        extra_authorities,
        extra_authority_count,
        manifest_hash: manifest.manifest_hash,
        declared_at: now,
        effective_at: 0,
    };
    let reading = observe(&ctx.accounts.covered_token_account, &clock);
    require!(
        declared.covers(&reading, &policy.holder, &policy.agent_address),
        CovanticError::GovernanceBaselineNotBoundToAccount
    );

    let baseline = &mut ctx.accounts.baseline;
    let is_new = baseline.policy_id == 0 && baseline.effective_at == 0;

    // A first declaration has no predecessor, and must say so. Seeding `prev_*`
    // with this declaration's own values and `now` is the natural thing to
    // write and exactly wrong: any `prev`-based fallback would then accept the
    // declaration as matured the instant it was written, silently disabling
    // the whole maturity delay.
    if is_new {
        baseline.clear_predecessor();
    } else {
        baseline.retain_as_predecessor();
    }

    baseline.policy_id = policy.policy_id;
    baseline.holder = policy.holder;
    baseline.token_owner = manifest.token_owner;
    baseline.expected_delegate = manifest.expected_delegate;
    baseline.expected_close_authority = manifest.expected_close_authority;
    baseline.program_upgrade_authority = None;
    baseline.controller = None;
    baseline.controller_min_threshold = 0;
    baseline.extra_authorities = extra_authorities;
    baseline.extra_authority_count = extra_authority_count;
    baseline.manifest_hash = manifest.manifest_hash;
    baseline.declared_at = now;
    baseline.effective_at = now
        .checked_add(GOVERNANCE_BASELINE_DELAY)
        .ok_or(CovanticError::MathOverflow)?;
    baseline.bump = ctx.bumps.baseline;

    emit!(GovernanceBaselineDeclared {
        policy_id: baseline.policy_id,
        holder: baseline.holder,
        token_owner: baseline.token_owner,
        extra_authority_count: baseline.extra_authority_count,
        manifest_hash: baseline.manifest_hash,
        declared_at: baseline.declared_at,
        effective_at: baseline.effective_at,
    });

    Ok(())
}

#[derive(Accounts)]
pub struct DeclareGovernanceBaseline<'info> {
    /// The policyholder. Nobody else may say who is allowed to control their
    /// agent — least of all the oracle, whose discretion this account exists
    /// to constrain.
    #[account(mut)]
    pub holder: Signer<'info>,

    #[account(seeds = [CONFIG_SEED], bump = config.bump)]
    pub config: Box<Account<'info, ProtocolConfig>>,

    #[account(
        seeds = [POLICY_SEED, policy.holder.as_ref(), &policy.policy_id.to_le_bytes()],
        bump = policy.bump,
        constraint = policy.holder == holder.key() @ CovanticError::UnauthorizedHolder,
    )]
    pub policy: Box<Account<'info, InsurancePolicy>>,

    /// The covered account the declaration is checked against, derived by
    /// address for the reason `checkpoint_authority` gives: a declaration may
    /// legitimately name an operator as the owner, and an ownership
    /// constraint would refuse to load the account in exactly that case.
    #[account(
        address = get_associated_token_address(&policy.agent_address, &usdc_mint.key())
            @ CovanticError::InvalidCoveredAccount,
    )]
    pub covered_token_account: Box<Account<'info, TokenAccount>>,

    #[account(constraint = usdc_mint.key() == config.usdc_mint @ CovanticError::InvalidTokenAccount)]
    pub usdc_mint: Box<Account<'info, Mint>>,

    #[account(
        init_if_needed,
        payer = holder,
        space = GovernanceBaseline::LEN,
        seeds = [GOVERNANCE_BASELINE_SEED, policy.key().as_ref()],
        bump,
    )]
    pub baseline: Box<Account<'info, GovernanceBaseline>>,

    pub system_program: Program<'info, System>,
}

File: packages/api/scripts/declare-governance-baseline.ts
/**
 * Declare — or refresh — the authority set that is legitimate for an agent.
 *
 * Usage:
 *   pnpm gov:declare --policy <id> --keypair keys/holder.json
 *   pnpm gov:declare --policy 7 --operator <pubkey> --operator <pubkey>
 *   pnpm gov:declare --policy 7 --delegate <pubkey> --close-authority <pubkey>
 *   pnpm gov:declare --policy 7 --extension-hash <64 hex chars>
 *
 * Why this is a holder-signed CLI and not something the oracle does for you:
 * the whole value of the declaration is that the *policyholder* made it. A
 * baseline the operator could write would put the operator back in charge of
 * the fact that is supposed to constrain them, and a governance claim proven
 * against it would prove nothing.
 *
 * The declaration matures an hour after it lands
 * (`GOVERNANCE_BASELINE_DELAY`). Until then a claim cannot be proven against
 * it — which is deliberate, and is what forces an attacker holding a stolen
 * holder key to pre-commit on chain, in public, well before the incident.
 *
 * The program reads the covered account while declaring and refuses a
 * declaration the account does not currently satisfy: the owner, any
 * delegate and any close authority must all be inside the declared set. A
 * declaration is a statement about the account as it stands, not a story.
 *
 * Refreshing keeps the whole previous declaration in `prev_*`, so rotating an
 * operator does not erase the record of what was legitimate yesterday.
 *
 * Program upgrade authorities and multisig controllers are not accepted: the
 * checkpoint reads a token account and cannot observe either, so the program
 * refuses to present them as covered. `--upgrade-authority`,
 * `--controller` and `--min-threshold` exit with an explanation.
 */

import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { config as loadDotenv } from 'dotenv';

loadDotenv({ path: resolve(import.meta.dirname, '../../../.env') });

import { AnchorProvider, Program, Wallet, type Idl } from '@coral-xyz/anchor';
import { Connection, Keypair, PublicKey, SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID, getAssociatedTokenAddressSync } from '@solana/spl-token';
import { governanceManifestCommitment, PDA_SEEDS, policyIdToBytes } from '@covantic/shared';

const REPO_ROOT = resolve(import.meta.dirname, '../../..');

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value || value.length === 0) throw new Error(`Missing required env var: ${name}`);
  return value;
}

function loadKeypair(path: string): Keypair {
  const abs = path.startsWith('/') ? path : resolve(REPO_ROOT, path);
  return Keypair.fromSecretKey(Uint8Array.from(JSON.parse(readFileSync(abs, 'utf-8')) as number[]));
}

/** Collect repeated `--flag value` pairs; single-valued flags take the last. */
function parseArgs(argv: string[]): Map<string, string[]> {
  const args = new Map<string, string[]>();
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token?.startsWith('--')) continue;
    const key = token.slice(2);
    const value = argv[i + 1];
    if (value === undefined || value.startsWith('--')) continue;
    args.set(key, [...(args.get(key) ?? []), value]);
    i += 1;
  }
  return args;
}

function optionalKey(args: Map<string, string[]>, name: string): PublicKey | null {
  const value = args.get(name)?.at(-1);
  return value ? new PublicKey(value) : null;
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));

  const policyId = args.get('policy')?.at(-1);
  if (!policyId) throw new Error('--policy <id> is required');

  for (const unsupported of ['upgrade-authority', 'controller', 'min-threshold']) {
    if (args.has(unsupported)) {
      throw new Error(
        `--${unsupported} is not accepted: the program cannot observe a program's upgrade ` +
          'authority or a multisig config from the covered token account, and refuses to ' +
          'record coverage it cannot settle.',
      );
    }
  }

  const keypairPath = args.get('keypair')?.at(-1) ?? requireEnv('ORACLE_KEYPAIR_PATH');
  const holder = loadKeypair(keypairPath);

  const connection = new Connection(requireEnv('SOLANA_RPC_URL'), 'confirmed');
  const provider = new AnchorProvider(connection, new Wallet(holder), {
    commitment: 'confirmed',
  });
  const idl = JSON.parse(
    readFileSync(resolve(REPO_ROOT, 'packages/anchor/target/idl/covantic.json'), 'utf-8'),
  ) as Idl;
  const program = new Program(idl, provider);

  const [config] = PublicKey.findProgramAddressSync(
    [Buffer.from(PDA_SEEDS.CONFIG)],
    program.programId,
  );
  const [policy] = PublicKey.findProgramAddressSync(
    [
      Buffer.from(PDA_SEEDS.POLICY),
      holder.publicKey.toBuffer(),
      Buffer.from(policyIdToBytes(BigInt(policyId))),
    ],
    program.programId,
  );
  const [baseline] = PublicKey.findProgramAddressSync(
    [Buffer.from(PDA_SEEDS.GOVERNANCE_BASELINE), policy.toBuffer()],
    program.programId,
  );

  const accounts = program.account as unknown as Record<
    string,
    { fetch: (a: PublicKey) => Promise<unknown> }
  >;
  const onChainPolicy = (await accounts.insurancePolicy!.fetch(policy)) as {
    agentAddress: PublicKey;
  };
  const onChainConfig = (await accounts.protocolConfig!.fetch(config)) as { usdcMint: PublicKey };
  const coveredTokenAccount = getAssociatedTokenAddressSync(
    onChainConfig.usdcMint,
    onChainPolicy.agentAddress,
  );

  // Defaults to the agent itself, which is the ordinary case: the agent owns
  // its own token accounts and nobody else may.
  const tokenOwner = optionalKey(args, 'token-owner') ?? onChainPolicy.agentAddress;
  const expectedDelegate = optionalKey(args, 'delegate');
  const expectedCloseAuthority = optionalKey(args, 'close-authority');
  const extraAuthorities = (args.get('operator') ?? []).map((k) => new PublicKey(k));
  if (extraAuthorities.length > 4) {
    throw new Error('At most 4 --operator addresses (MAX_GOVERNANCE_EXTRA_AUTHORITIES)');
  }
  const extensionHex = args.get('extension-hash')?.at(-1);
  const extensionHash = extensionHex
    ? Uint8Array.from(Buffer.from(extensionHex, 'hex'))
    : undefined;
  if (extensionHash && extensionHash.length !== 32) {
    throw new Error('--extension-hash must be 32 bytes of hex');
  }

  // Commits to the declaration as written — every field of it — plus the
  // digest of any richer off-chain document. The program cannot decode a
  // Squads config or an allowed-signer list, so the richer statement lives
  // off chain and this hash is what makes it permanently falsifiable.
  const manifestHash = governanceManifestCommitment({
    tokenOwner: tokenOwner.toBytes(),
    expectedDelegate: expectedDelegate?.toBytes() ?? null,
    expectedCloseAuthority: expectedCloseAuthority?.toBytes() ?? null,
    programUpgradeAuthority: null,
    controller: null,
    controllerMinThreshold: 0,
    extraAuthorities: extraAuthorities.map((k) => k.toBytes()),
    extensionHash,
  });

  const manifest = {
    tokenOwner,
    expectedDelegate,
    expectedCloseAuthority,
    programUpgradeAuthority: null,
    controller: null,
    controllerMinThreshold: 0,
    extraAuthorities,
    manifestHash: Array.from(manifestHash),
  };

  const signature = await (
    program.methods as unknown as Record<
      string,
      (m: unknown) => {
        accounts: (a: Record<string, PublicKey>) => { rpc: () => Promise<string> };
      }
    >
  ).declareGovernanceBaseline!(manifest)
    .accounts({
      holder: holder.publicKey,
      config,
      policy,
      coveredTokenAccount,
      usdcMint: onChainConfig.usdcMint,
      baseline,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: SystemProgram.programId,
    })
    .rpc();

  // Read it back rather than computing it — see the note in
  // `declare-agent-mandate.ts`. A `devnet-fast-lock` build compresses the
  // on-chain delay, and a locally-derived figure would tell a policyholder to
  // wait for something that has already happened.
  const onChain = (await accounts.governanceBaseline!.fetch(baseline)) as {
    effectiveAt: { toNumber(): number };
  };
  const effectiveAt = new Date(onChain.effectiveAt.toNumber() * 1000);
  process.stdout.write(
    [
      `Declared governance baseline for policy ${policyId}`,
      `  baseline PDA : ${baseline.toBase58()}`,
      `  token owner  : ${tokenOwner.toBase58()}`,
      `  delegate     : ${expectedDelegate?.toBase58() ?? '(none)'}`,
      `  close auth.  : ${expectedCloseAuthority?.toBase58() ?? '(none)'}`,
      `  operators    : ${extraAuthorities.map((k) => k.toBase58()).join(', ') || '(none)'}`,
      `  manifest hash: ${Buffer.from(manifestHash).toString('hex')}`,
      `  signature    : ${signature}`,
      '',
      `Usable as proof from ${effectiveAt.toISOString()} — a claim filed before then`,
      'cannot be proven against it. That delay is the mechanism, not a formality.',
      '',
    ].join('\n'),
  );
}

main().catch((err) => {
  process.stderr.write(`${err instanceof Error ? err.stack : String(err)}\n`);
  process.exit(1);
});
